src/connector/src/sink/doris_starrocks_connector.rs

```diff
@@ -386,6 +386,64 @@ impl InserterInner {
     }
 }
 
+enum StreamLoadResponse {
+    BeRequest(Request),
+    HttpResponse(Response),
+}
+
+/// Send the request and handle redirection if any.
+/// The reason we handle the redirection manually is that if we let `reqwest` handle the redirection
+/// automatically, it will remove sensitive headers (such as Authorization) during the redirection,
+/// and there's no way to prevent this behavior.
+/// Please note, the FE address that user specified might be a FE follower not the leader, in this case,
+/// the follower FE will redirect request to leader FE and then to BE.
+async fn send_stream_load_request(
+    client: Client,
+    mut request: Request,
+    fe_host: &str,
+) -> Result<StreamLoadResponse> {
+    // possible redirection paths:
+    // RW <-> follower FE -> leader FE -> BE
+    // RW <-> leader FE -> BE
+    // RW <-> leader FE
+    for _ in 0..2 {
+        let original_http_port = request.url().port();
+        let mut request_for_redirection = request
+            .try_clone()
+            .ok_or_else(|| SinkError::DorisStarrocksConnect(anyhow!("Can't clone request")))?;
+        let resp = client.execute(request).await.map_err(|err| {
+            SinkError::DorisStarrocksConnect(
+                anyhow!(err).context("sending stream load request failed"),
+            )
+        })?;
+        let be_url = try_get_be_url(&resp, fe_host)?;
+        match be_url {
+            Some(be_url) => {
+                // we used an unconventional method to detect if we are currently redirecting to FE leader, i.e.,
+                // by comparing the port of the redirected url with that of the original request, if they are same, we consider
+                // this is a FE address. Because in practice, no one would deploy their `StarRocks` cluster with the same
+                // http port for both FE and BE. However, this is a potentially problematic assumption,
+                // we may investigate a better way to do this. For example, we could use the `show backends` command to check
+                // if the host of the redirected url is in the list. However, `show backends` requires
+                // the system-level privilege, which could break the backward compatibility.
+                let redirected_port = be_url.port();
```

Comment:
In case where user deploy the cluster with the same HTTP port for both FE and BE, a load request with an empty body will be sent to BE. And then SR will respond with an error: `all partitions have no load data`, causing the sink to fail repeatedly.

Reply:
Is it possible to ignore the error (log only) when getting the fe address to avoid repeated failures?

Reply:
It may be expected to let the sink fail, as we do not support this unusual deployment topology (where FE's http-port is the same as the BE's http-port) after this PR, though it is rare in real production cases.

Reply:
what I mean is, when this error `（all partitions have no load data）` occurs in this method, can we determine the error and ignore it (or just hit a log) to avoid rebuilding the sink

Reply:
https://github.com/risingwavelabs/risingwave/blob/39c14adc12511613c05d417d80838aa17d35a6e9/src/connector/src/sink/doris_starrocks_connector.rs#L599-L608

`expect a be request but got a response` this error will report in this case. I just tested sending BE an empty request through `/api/transaction/load` which does not cause BE to respond with an error, but the subsequent request to `/api/transaction/prepare` will respond with `all partitions have no load data`.

Reply:
After looking at the code, `send_stream_load_request` seems to be called in two places, the problem here will only happen when load is called, when the situation described above is happening, we will get a http resp and report an error, but actually we have already gotten the be_resp in the previous step, but we think it is fe_resp, is it possible to save it and try to send it using the be_resp(fe_resp), do you think it's possible?

Reply:
I just tested that sending BE a normal `load` right after an empty `load` allows the whole transaction to succeed without any errors and without data loss, so I think the idea you mentioned is reasonable. Let me fix it.

Reply:
@xxhZs PTAL, BTW, we wrote a lot of code related to handling manual redirection, and it is all because the `reqwest` removes the sensitive headers during the redirection. This is annoying and makes our code too complicated.

Reply:
When I am searching for whether there are other rust http clients that can keep the sensitive headers after redirection, I find that stripping auth headers are by design for HTTP for security reasons: https://community.cloudflare.com/t/prevent-dropping-of-authorization-header-on-redirect/677517 :(